db/src/tx.rs

```diff
@@ -695,63 +698,45 @@ impl<'conn, 'a, W> Tx<'conn, 'a, W> where W: TransactWatcher {
         // We need to ensure that callers can't blindly transact entities that haven't been
         // allocated by this store.
 
+        let errors = tx_checking::type_disagreements(&aev_trie);
+        if !errors.is_empty() {
+            bail!(ErrorKind::SchemaConstraintViolation(errors::SchemaConstraintViolation::TypeDisagreements { conflicting_datoms: errors }));
+        }
+
+        let errors = tx_checking::cardinality_conflicts(&aev_trie);
```

Comment:
Same here. You know there are two assertions because you encounter the other one while you're inserting…